src/passa/lockers.py

```diff
@@ -0,0 +1,180 @@
+# -*- coding=utf-8 -*-
+
+from __future__ import absolute_import, unicode_literals
+
+import itertools
+
+import plette
+import requirementslib
+import resolvelib
+import vistir
+
+from .caches import HashCache
+from .hashes import get_hashes
+from .metadata import set_metadata
+from .providers import BasicProvider, EagerUpgradeProvider, PinReuseProvider
+from .reporters import StdOutReporter
+from .traces import trace_graph
+from .utils import identify_requirment
+
+
+def _get_requirements(model, section_name):
+    """Produce a mapping of identifier: requirement from the section.
+    """
+    if not model:
+        return {}
+    return {identify_requirment(r): r for r in (
+        requirementslib.Requirement.from_pipfile(name, package._data)
+        for name, package in model.get(section_name, {}).items()
+    )}
+
+
+def _iter_derived_entries(state, traces, names):
+    """Produce a mapping containing all candidates derived from `names`.
+
+    `name` should provide a collection of requirement identifications from
+    a section (i.e. `packages` or `dev-packages`). This function uses `trace`
+    to filter out candidates in the state that are present because of an entry
+    in that collection.
+    """
+    if not names:
+        return
+    names = set(names)
+    for name, requirement in state.mapping.items():
+        routes = {trace[1] for trace in traces[name] if len(trace) > 1}
+        if name not in names and not (names & routes):
+            continue
+        yield (
+            requirement.normalized_name,
+            next(iter(requirement.as_pipfile().values()))
+        )
+
+
+class AbstractLocker(object):
```

Comment:
I actually love that this is called a locker, not sure if it translates properly but it's also the equivalent of where you lock things up for safe keeping